### PR TITLE
Visual update for computeparticles

### PR DESCRIPTION
**Pull request by Voultapher**
Instead of using a small amount of large textured particles, use a large
amount of small monochrome particles.

geometryshader: added togglePipeline functionality for wireframe
rendering

vulkanexamplebase: added functionality to create an only device visible
buffer and updating it via a temporary staging buffer.

### CODE REVIEW

**Comment by SaschaWillems** [comment]
Thx for your pull request :+1: . I see that it does not only contain change the compute particle example but also to other samples as well as the base class, so I'll have to check first before (if) I merge. 

Also please note that the example is supposed to run with a nice framerate on lower end mobile devices, so the particle count is supposed to be moderate.

My spare time is a bit limited right now, but I hope to check your changes sometime this week.

**Comment by Voultapher** [comment]
It runs at 200fps on a 780ti, the expensive part is by far the drawcall executed by the fragment shader. The visual appearance is mainly paired to 2 values. For one the number of particles and the opacity of such. I saw some of your multi threading examples go as low as 30fps on a desktop. So I figured 200fps would be enough of a headroom for weaker devices and still be visually impressive. If you think that's to expensive I can tweak it.

**Comment by SaschaWillems** [comment]
>  I saw some of your multi threading examples go as low as 30fps on a desktop. So I figured 200fps would be enough of a headroom for weaker devices and still be visually impressive.

200 fps sound reasonable :racehorse: . I'm testing all my examples on an NVIDIA Shield TV too, which has different performance characteristics than some desktop GPUs and just want to make sure it runs reasonably well there too. 

Will check out your PR soon.

File: base/vulkanexamplebase.cpp [hunk 1]
```diff
@@ -18,7 +21,7 @@ VkResult VulkanExampleBase::createInstance(bool enableValidation)
 	appInfo.pEngineName = name.c_str();
 	// Temporary workaround for drivers not supporting SDK 1.0.3 upon launch
 	// todo : Use VK_API_VERSION
-	appInfo.apiVersion = VK_MAKE_VERSION(1, 0, 2);
+	appInfo.apiVersion = VK_MAKE_VERSION(1, 0, 5);
```
Comment (1ace) [inline base/vulkanexamplebase.cpp]: Why did you do this? I could be mistaken, but I don't think this can affect the code, but it will affect at runtime the ability for users to link the app against older (but compatible) drivers.

**Comment by 1ace** [comment]
>  see that it does not only contain change the compute particle example but also to other samples as well as the base class

Just my 2 cents, but it's usually considered best practice to have that kind of unrelated changes in separate commits to make it easier to review (and separate pull-requests to make it easier to merge).

For instance I've reviewed parts of this PR, and what I've seen looks good *(with one exception, but that could just be me being wrong :)*, but I can't (yet) say anything about the whole.

BTW, the fact you wrote 3 different commit messages into that one commit should've been a hint :wink:

File: base/vulkanexamplebase.cpp (hunk 1 above)
Comment (Voultapher) [inline base/vulkanexamplebase.cpp]: I changed it trying to solve a driver bug. Forgot to change it back.

**Comment by SaschaWillems** [comment]
@Voultapher : Can you please do a new pull request that only contains changes to the compute particle example? I don't want the other changes added to my repository, and afaik I can't just merge parts of a PR.

Thx

**Comment by 1ace** [comment]
> afaik I can't just merge parts of a PR.

That's true, but he could also amend his commit to only contain the relevant stuff and force-push that to his fork, which will update this PR :)

**Comment by Voultapher** [comment]
When going through the code I found that naming the vulkan return value variable "err" would imply that it is an error message/ information and the norm is VK_SUCCESS which imo is no error. I think naming it sth like "res" and changing the asserts to "assert(res == VK_SUCCESS);" would clarify the code.

Concerning the PR I'll try a force push.

**Comment by SaschaWillems** [comment]
> When going through the code I found that naming the vulkan return value variable "err" would imply that it is an error message/ information and the norm is VK_SUCCESS which imo is no error. I think naming it sth like "res" and changing the asserts to "assert(res == VK_SUCCESS);" would clarify the code.

Yes, that's true. Err is not a proper naming, that's why I recently introduced vkTools::checkResult. I'm going to replace all asserts with this function if I find some time.

> Concerning the PR I'll try a force push.

Thx in advance! I'd really like to have your compute shader in the repository :smile:

**Comment by Voultapher** [comment]
That should do it ;D

**Comment by 1ace** [comment]
No, you added 4 more commits instead of amending the commit you already had.

The way you can fix this now is to do an interactive [`rebase`](https://git-scm.com/docs/git-rebase) and squash all those commits.
If it's too complicated for you, you can always just close this PR and make a new one.

Try to also pay attention to your commit messages: make sure they make sense, and use full sentences.

**Comment by 1ace** [comment]
Looking at the current PR, it still contains a lot of completely unrelated changes. Please do try and keep your commits coherent and organized.

**Comment by Voultapher** [comment]
As far as I can tell, the changes in the last commit are the bare minimum. Please be more specific about the unrelated changes.

File: base/vulkanexamplebase.cpp [hunk 2]
```diff
@@ -40,6 +40,7 @@ VkResult VulkanExampleBase::createInstance(bool enableValidation)
 		if (enableValidation)
 		{
 			enabledExtensions.push_back(VK_EXT_DEBUG_REPORT_EXTENSION_NAME);
+
```
Comment (1ace) [inline base/vulkanexamplebase.cpp]: Unnecessary empty line added

File: base/vulkanexamplebase.cpp [hunk 3]
```diff
@@ -268,7 +360,7 @@ VkBool32 VulkanExampleBase::createBuffer(
 	}
 	err = vkBindBufferMemory(device, *buffer, *memory, 0);
 	assert(!err);
-	return true;
+	return VK_TRUE;
```
Comment (1ace) [inline base/vulkanexamplebase.cpp]: Unrelated fix

File: base/vulkanexamplebase.cpp [hunk 4]
```diff
@@ -331,7 +423,10 @@ void VulkanExampleBase::renderLoop()
 				DispatchMessage(&msg);
 			}
 		}
+
+		// render
 		render();
+
```
Comment (1ace) [inline base/vulkanexamplebase.cpp]: Unrelated

File: computeparticles/computeparticles.cpp [hunk 5]
```diff
@@ -58,7 +59,7 @@ class VulkanExample : public VulkanExampleBase
 		float deltaT;
 		float destX;
 		float destY;
-		int32_t particleCount = PARTICLE_COUNT;
+		uint32_t particleCount = PARTICLE_COUNT;
```
Comment (1ace) [inline computeparticles/computeparticles.cpp]: Unrelated fix

File: computeparticles/computeparticles.cpp [hunk 6]
```diff
@@ -68,19 +69,17 @@ class VulkanExample : public VulkanExampleBase
...
 	VkPipelineLayout pipelineLayout;
-	VkDescriptorSet descriptorSetPostCompute;
 	VkDescriptorSetLayout descriptorSetLayout;
 
 	VulkanExample() : VulkanExampleBase(ENABLE_VALIDATION)
 	{
-		width = 1280;
-		height = 720;
+		width = 1920;
+		height = 1080;
```
Comment (1ace) [inline computeparticles/computeparticles.cpp]: Unrelated change

File: computeparticles/computeparticles.cpp [hunk 7]
```diff
@@ -129,7 +117,7 @@ class VulkanExample : public VulkanExampleBase
 
 		VkClearValue clearValues[2];
 		clearValues[0].color = defaultClearColor;
-		clearValues[1].depthStencil = { 1.0f, 0 };
+		clearValues[1].depthStencil = { 1.f, 0 };
```
Comment (1ace) [inline computeparticles/computeparticles.cpp]: Unnecessary change

File: computeparticles/computeparticles.cpp [hunk 8]
```diff
@@ -142,7 +130,7 @@ class VulkanExample : public VulkanExampleBase
 
 		VkResult err;
 
-		for (int32_t i = 0; i < drawCmdBuffers.size(); ++i)
+		for (size_t i = 0; i < drawCmdBuffers.size(); ++i)
```
Comment (1ace) [inline computeparticles/computeparticles.cpp]: Unrelated fix (which btw is already in a PR waiting to be merged)

File: computeparticles/computeparticles.cpp [hunk 9]
```diff
@@ -151,7 +139,7 @@ class VulkanExample : public VulkanExampleBase
 			assert(!err);
 
 			// Buffer memory barrier to make sure that compute shader
-			// writes are finished before using the storage buffer 
+			// writes are finished before using the storage buffer
```
Comment (1ace) [inline computeparticles/computeparticles.cpp]: Unrelated change

File: computeparticles/computeparticles.cpp [hunk 10]
```diff
@@ -183,15 +171,9 @@ class VulkanExample : public VulkanExampleBase
 				);
 			vkCmdSetViewport(drawCmdBuffers[i], 0, 1, &viewport);
 
-			VkRect2D scissor = vkTools::initializers::rect2D(
-				width,
-				height,
-				0,
-				0
-				);
+			VkRect2D scissor = vkTools::initializers::rect2D(width, height, 0, 0);
```
Comment (1ace) [inline computeparticles/computeparticles.cpp]: Unrelated change

File: computeparticles/computeparticles.cpp [hunk 11]
```diff
@@ -262,93 +244,26 @@ class VulkanExample : public VulkanExampleBase
...
+
+		std::mt19937 rGenerator;
+		std::uniform_real_distribution<float> rDistribution(-1.f, 1.f);
 
 		// Initial particle positions
-		std::vector<Particle> particleBuffer;
-		for (int i = 0; i < PARTICLE_COUNT; ++i)
+		std::vector<Particle> particleBuffer(PARTICLE_COUNT);
+
+		for (auto& element : particleBuffer)
```
Comment (1ace) [inline computeparticles/computeparticles.cpp]: Unrelated change (although I approve)

File: computeparticles/computeparticles.cpp [hunk 12]
```diff
@@ -562,7 +432,7 @@ class VulkanExample : public VulkanExampleBase
 		blendAttachmentState.dstAlphaBlendFactor = VK_BLEND_FACTOR_DST_ALPHA;
 
 		err = vkCreateGraphicsPipelines(device, pipelineCache, 1, &pipelineCreateInfo, nullptr, &pipelines.postCompute);
-		assert(!err);
+		assert(err == VK_SUCCESS);
```
Comment (1ace) [inline computeparticles/computeparticles.cpp]: Unrelated change

File: computeparticles/computeparticles.cpp [hunk 13]
```diff
@@ -681,8 +558,7 @@ class VulkanExample : public VulkanExampleBase
 		vkGetPhysicalDeviceQueueFamilyProperties(physicalDevice, &queueCount, NULL);
 		assert(queueCount >= 1);
 
-		std::vector<VkQueueFamilyProperties> queueProps;
-		queueProps.resize(queueCount);
+		std::vector<VkQueueFamilyProperties> queueProps(queueCount);
```
Comment (1ace) [inline computeparticles/computeparticles.cpp]: Unrelated change

File: data/shaders/computeparticles/particle.comp [hunk 14]
```diff
@@ -5,62 +5,73 @@
...
+layout(std140, binding = 0) buffer Pos
 {
-   Particle particles[ ];
+	Particle particles[ ];
 };
 
 layout (local_size_x = 16, local_size_y = 16) in;
 
-layout (binding = 1) uniform UBO 
+layout (binding = 1) uniform UBO
```
Comment (1ace) [inline data/shaders/computeparticles/particle.comp]: Last 3: unrelated changes

File: data/shaders/computeparticles/particle.comp [hunk 15]
```diff
@@ -5,62 +5,73 @@
...
 {
 	float deltaT;
 	float destX;
 	float destY;
-	int particleCount;
+	uint particleCount;
```
Comment (1ace) [inline data/shaders/computeparticles/particle.comp]: Unrelated fix

File: data/shaders/computeparticles/particle.comp [hunk 16]
```diff
@@ -5,62 +5,73 @@
...
+	vec2 delta = attractPos - pos;
 	const float damp = 0.5;
-    float dDampedDot = dot(delta, delta) + damp;
-    float invDist = 1.0f / sqrt(dDampedDot);
-    float invDistCubed = invDist*invDist*invDist;
-    return delta * invDistCubed * 0.00035;
+	float dDampedDot = dot(delta, delta) + damp;
+	float invDist = 1.0f / sqrt(dDampedDot);
+	float invDistCubed = invDist*invDist*invDist;
+	return delta * invDistCubed * 0.0035;
```
Comment (1ace) [inline data/shaders/computeparticles/particle.comp]: Last 4 lines: unrelated change (except for the value change)

File: data/shaders/computeparticles/particle.comp [hunk 17]
```diff
@@ -5,62 +5,73 @@
...
-		
-    // Move by velocity
-    vPos += vVel * ubo.deltaT;
-	
-	if ((vPos.x < -1.0) || (vPos.x > 1.0))
-		vVel.x -= vVel.x;
-
-    // Write back
-    particles[index].pos.xyz = vPos;
-    particles[index].vel.xyz = vVel;
```
Comment (1ace) [inline data/shaders/computeparticles/particle.comp]: I suspect you didn't change anything from `main()`, but your reformatting confused git: unrelated change

**Comment by 1ace** [comment]
I skipped a few, but it should give you a good idea of what you can improve :+1:

**Comment by Voultapher** [comment]
Ok, thx

File: computeparticles/computeparticles.cpp (hunk 5 above)
Comment (Voultapher) [inline computeparticles/computeparticles.cpp]: later used in paticle.comp and compared to a uint, particleCount should never be negative.

File: computeparticles/computeparticles.cpp (hunk 7 above)
Comment (Voultapher) [inline computeparticles/computeparticles.cpp]: The 0 is unnecessary.

**Comment by 1ace** [comment]
A couple rules of thumb:
- Formatting changes are frowned upon: if your change doesn't improve the code behaviour and doesn't make the the code easier to read, don't do it.
- Keep unrelated change separate (different commits, different pull requests when possible)
- Always check the outstanding PRs for a similar change. Don't submit yet another change on the same piece of code before the previous one has been either accepted or rejected.

To be clear, most of your "unrelated changes" are things that I think could be useful, but don't belong in this PR.

File: computeparticles/computeparticles.cpp (hunk 5 above)
Comment (1ace) [inline computeparticles/computeparticles.cpp]: Yes, which is why I called it a "fix", but it's not related to the current change. Send it as a new PR if you want :]

File: computeparticles/computeparticles.cpp (hunk 7 above)
Comment (1ace) [inline computeparticles/computeparticles.cpp]: This one is a matter of taste, and therefore falls in the "formatting change" category I mentioned earlier. It doesn't make the code objectively better, so it should be kept the way the original author wrote it.
